### PR TITLE
LOOKDEVX-470 : Initial support for export.

### PR DESCRIPTION
**Pull request by feldstj**

### CODE REVIEW

File: resources/LookDev/defaultLook.mtlx [hunk 1]
```diff
@@ -11,5 +11,5 @@
   <look name="defaultLook">
     <materialassign name="defaultMaterialAssign" collection="defaultCollection" material="defaultMaterial"/>
   </look>
-  <lookgroup name="defaultLookGroup" looks="defaultLook" activeLook="defaultLook" />
+  <lookgroup name="defaultLookGroup" looks="defaultLook" enabled="defaultLook" />
```
Comment (feldstj) [inline resources/LookDev/defaultLook.mtlx:14]: defaultLook.mtlx fix

File: resources/Materials/TestSuite/stdlib/looks/looks.mtlx [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0"?>
```
Comment (feldstj) [inline resources/Materials/TestSuite/stdlib/looks/looks.mtlx:1]: Needed for test case.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 3]
```diff
@@ -1650,5 +1655,102 @@ void RtFileIo::writePrim(std::ostream& stream, const RtPath& primPath, const RtW
     writeToXmlStream(document, stream);
 }
 
+void mergeLooks(DocumentPtr document)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp:1658]: Function that merges the look information into a single look.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 4]
```diff
@@ -1650,5 +1655,102 @@ void RtFileIo::writePrim(std::ostream& stream, const RtPath& primPath, const RtW
...
+        }
+    }
+    // Delete all the looks from the document that are in lookNames
+    for (const std::string& lookName : lookNames)
+    {
+        document->removeChild(lookName);
+    }
 }
 
+void RtFileIo::exportDocument(std::ostream& stream, const RtExportOptions* options)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp:1716]: Export routines.

File: source/MaterialXRuntime/RtFileIo.h [hunk 5]
```diff
@@ -82,6 +82,18 @@ class RtWriteOptions
     unsigned int desiredMinorVersion;
 };
 
+/// @class RtExportOptions
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.h:85]: RtExportOptions.

File: source/MaterialXTest/MaterialXRuntime/Runtime.cpp [hunk 6]
```diff
@@ -2719,3 +2719,54 @@ TEST_CASE("Runtime: duplicate name", "[runtime]")
     REQUIRE(graph1.getOutputSocket(ADD5));
     REQUIRE(duplicateCount(ADD5) == 1);
 }
+
+
+TEST_CASE("Export", "[runtime]")
```
Comment (feldstj) [inline source/MaterialXTest/MaterialXRuntime/Runtime.cpp:2724]: Simple test case.

**Review by bernardkwok — CHANGES_REQUESTED** [review]
Some logic comments and also that we require export at the core level and not inside runtime
for general usage.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 7]
```diff
@@ -1650,5 +1655,102 @@ void RtFileIo::writePrim(std::ostream& stream, const RtPath& primPath, const RtW
...
+    }
+    // Delete the non-mainLookGroup lookgroups
+    for (const std::string& lookgroupName : lookgroupNames)
+    {
+        document->removeChild(lookgroupName);
+    }
+    // Append looks which are not a part of a lookgroup to the mainLookGroup
+    for (LookPtr look : document->getLooks())
+    {
+        if (looksInLookGroup.count(look->getName()) == 0)
```
[the remark this answers is not in the document]
Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp:1691]: Ok, I'll add the stray looks first. I actually like == 0 better to be honest, I think it's more clear.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 8]
```diff
@@ -1650,5 +1655,102 @@ void RtFileIo::writePrim(std::ostream& stream, const RtPath& primPath, const RtW
...
+    LookPtr mainLook = mainLookGroup->combineLooks();
+    // Delete the mainLookGroup
+    document->removeChild(mainLookGroup->getName());
+    // Append look names that don't belong to the mainLook to lookNames
+    std::vector<std::string> lookNames;
+    for (LookPtr look : document->getLooks())
+    {
+        if (look != mainLook)
+        {
+            lookNames.push_back(look->getName());
```
[the remark this answers is not in the document]
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp:1706]: Guess you could iterate in reverse. If not guess leave it as is.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp:1691]: Think we use the "!" convention so just mentiong to stick to whatever the convention is.

[the remark this answers is not in the document]
Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp:1691]: I just did a search in the MaterialX repo for .count() and it seems that both are used about equally.

File: source/MaterialXCore/Document.cpp [hunk 9]
```diff
@@ -153,6 +153,75 @@ void Document::initialize()
...
+        mainLookGroup->appendLookGroup(lookgroup);
+
+        // Delete the non-mainLookGroup lookgroups
+        removeChild(lookgroupName);
+    }
+    // Combine the mainLookGroup into a mainLook
+    LookPtr mainLook = mainLookGroup->combineLooks();
+    // Delete the mainLookGroup
+    removeChild(mainLookGroup->getName());
+    // Append look names that don't belong to the mainLook to lookNames
```
Comment (bernardkwok) [inline source/MaterialXCore/Document.cpp:209]: "append the looks which are not the main look."

File: source/MaterialXFormat/XmlIo.cpp [hunk 10]
```diff
@@ -362,6 +372,32 @@ string writeToXmlString(DocumentPtr doc, const XmlWriteOptions* writeOptions)
     return stream.str();
 }
 
+static void mergeLooks(DocumentPtr doc, const XmlExportOptions* exportOptions)
```
Comment (bernardkwok) [inline source/MaterialXFormat/XmlIo.cpp:375]: We should move this to the anonymous namespace at the top and make non-static.

**Review by bernardkwok — APPROVED** [review]
Looks good.

**Pull request merged**